components/layout_2020/query.rs

```diff
@@ -509,8 +514,415 @@ fn is_eligible_parent(fragment: &BoxFragment) -> bool {
+                }
+
+                // Step 3: If node is not being rendered, then return items. For the purpose of this step,
+                // the following elements must act as described if the computed value of the 'display'
+                // property is not 'none':
+                let display = style.get_box().display;
+                if display == Display::None {
+                    match element.type_id() {
+                        // Even if set to Display::None, Option/OptGroup elements need to
+                        // be rendered.
+                        LayoutNodeType::Element(LayoutElementType::HTMLOptGroupElement) |
+                        LayoutNodeType::Element(LayoutElementType::HTMLOptionElement) => {},
+                        _ => {
+                            return items;
+                        },
+                    }
+                }
+
+                let text_content = node.to_threadsafe().node_text_content();
+
+                let white_space_collapse = style.clone_white_space_collapse();
+                let preserve_whitespace = white_space_collapse == WhiteSpaceCollapseValue::Preserve;
+                // Now we need to decide on whether to remove beginning white space or not, this
+                // is mainly decided by the elements we rendered before, but may be overwritten by the white-space
+                // property.
+                let trim_beginning_white_space = !preserve_whitespace &&
+                    (state.may_start_with_whitespace || display == Display::InlineBlock);
+                let with_white_space_rules_applied: WhitespaceCollapse<_> = WhitespaceCollapse::new(
+                    text_content.chars(),
+                    white_space_collapse,
+                    trim_beginning_white_space,
+                );
+
+                // Step 4: If node is a Text node, then for each CSS text box produced by node, in
+                // content order, compute the text of the box after application of the CSS
+                // 'white-space' processing rules and 'text-transform' rules, set items to the list
+                // of the resulting strings, and return items. The CSS 'white-space' processing
+                // rules are slightly modified: collapsible spaces at the end of lines are always
+                // collapsed, but they are only removed if the line is the last line of the block,
+                // or it ends with a br element. Soft hyphens should be preserved.
+                let transformed_text = TextTransformation::new(
+                    with_white_space_rules_applied,
+                    style.clone_text_transform().case(),
+                )
+                .collect::<String>();
+
+                let is_preformatted_element =
+                    white_space_collapse == WhiteSpaceCollapseValue::Preserve;
+
+                let is_final_character_whitespace = transformed_text
+                    .chars()
+                    .last()
```

Comment:
Minor thing, but `.chars().last()` is O(n) since it needs to to iterate over all the previous characters, you could use [`next_back()`](https://doc.rust-lang.org/std/iter/trait.DoubleEndedIterator.html#tymethod.next_back) for O(1) instead.

Reply:
Ah thanks! I just changed it to use `next_back` instead.